### DISCUSSION TITLE
Can't use #[Lazy] or ->lazy() and #[Layout]

### DISCUSSION THREAD

**Discussion by adriansio** [livewire/livewire#6907] [7. Bugs (archived)]
### Livewire version

3.0.5

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

```
<?php

namespace App\Livewire;

use Livewire\Attributes\Lazy;
use Livewire\Component;

#[Lazy]
class TestLayout extends Component
{
    public function render()
    {
        return view('livewire.test-layout')
                     ->layout('layouts.admin'); // <-- the custom layout works only if I remove #[Lazy] or ->lazy()
    }

    public function mount()
    {
        sleep(1);
    }
}
```

### Please confirm (incomplete submissions will not be addressed)

- [x] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by gazzoy** [comment]
+1

I mentioned about this on #7107, but no response.

**Comment by lrljoe** [comment]
By changing the layout, it'll involve a full reload anyway ideally!

Lazy Loading a Component, and changing the layout, to my mind anyway doesn't seem like a valid approach

What are you trying to achieve with this?

What happens if you use the layout attribute rather than the method on the view?

**Reply by adriansio** [reply]
Just a component having a custom layout, which also needs to be loaded lazy. I don't see any issue with it.

**Reply by lrljoe** [reply]
A Layout file is for **layout**.  It's the base layer for your website, where you add the relevant includes, structure etc for your user to experience.

You cannot lazily swap between layouts.  A **layout** may contain different approaches, different underlying CSS/JS includes etc.  For example, your "guest" layout may not even include the Livewire core JS includes, but your "user" layout does.

I think you need to really rethink your approach if you're trying to swap between different layouts on the fly, the Laravel core documentation explains very clearly how Layouts behave.

A "component" being **wrapped** in a different parent Blade Component is different to a Layout.

**Reply by gazzoy** [reply]
@lrljoe I believe that adriansio does not want to swap a layout, but simply wants to use a custom layout for a fullpage-component(e.g., `layouts.admin` or `layouts.custom`). Which seems normal to me.

**Reply by lrljoe** [reply]
> @lrljoe I believe that adriansio does not want to swap a layout, but simply wants to use a custom layout for a fullpage-component(e.g., `layouts.admin` or `layouts.custom`). Which seems normal to me.

Now that makes more sense!  A full page component, lazily loaded.  Not swapping layouts midway etc.

I know the core team are aware of this bug, I had a quick poke, and there's a few bits that'd need adjusting to get it to work.

Personally, I've yet to end up with a use case for a full page component, lazily loaded

**Comment by gazzoy** [comment]
@lrljoe Many thanks for supporting this. I will wait for the fix then!

FYI, I've got a same problem on this. In my case, I tried following code, but no luck. 

web.php
```
Route::get('/do-something', DoSomething::class)->lazy();
```

DoSomething.php
```
#[Layout('layouts.admin')] // this does NOT work
class DoSomething extends Component
{
...
    public function placeholder(array $params = [])
    {
        return view('livewire.placeholders.skeleton', $params);
        // return view('livewire.placeholders.skeleton', $params)->layout("layouts.admin"); // this does NOT work either
    }
...
}
```

**Comment by lrljoe** [comment]
Noting link to other (closed) discussion: #7107 

It took me a while to replicate the issue, purely as I don't tend to lazily load full page components at all.  I know the core team are aware of the issue as I discussed with them.

Did a few tweaks, but it wasn't going to be a swift fix, and duty called sadly.

**Reply by adriansio** [reply]
@lrljoe Indeed it's what @gazzoy said. Basically I like to show a 'loading spinner' in every component, full page or not, as I want to load them instantly and then make slow calls once the page has loaded. This is a great case for the lazy feature in Livewire but if I am using a custom layout, this doesn't work.

That's probably why you haven't came across the use case, probably you aren't showing a loading placeholder on full page components.

**Comment by DanielSpravtsev** [comment]
Unfortunately, it is too deep bug, can't find a solution. Help would be appreciated, guys @calebporzio @joshhanley <3

**Reply by gazzoy** [reply]
@DanielSpravtsev Thanks for taking care of this. Will be waiting for the GOD's replies 👀

**Comment by joshhanley** [comment]
So I have a fair idea what is causing this.

Basically what is happening currently is when the lazy component placeholder is being loaded, it is just loading the default layout `resources/views/components/app.blade.php` as the component isn't actually being processed when this happens (only the placeholder method).

As such, when the lazy component actually loads, the default layout has already been loaded, so the component doesn't try to reload a layout as that will cause things to break. So the component just refreshes it's self and it's HTML and nothing else.

Naturally, this isn't ideal, as the lazy component placeholder load request should probably check the component to see if there are any `#[Layout()]` attributes available and make use of that.

The harder issue is if the original way of defining the layout, in the render method, is used, by calling `view()->layout()`. The problem with this is that the render method will actually need to be run, for us to be able to resolve the layout from the view.

But we can't do that with a lazy component, as none of the component's lifecycle has run (like mount), so there may be missing data which will cause the render method to throw an error.

**Short term work around solution**

As a work around, I wouldn't use full page lazy components. Instead I would just use a normal full page component, and if you need parts of that to be lazy (even if it's all of the HTML), then I would create a nested component, make it lazy and wrap up anything inside of there that you need to be deferred. Or you could just use a boolean property on your main component and `wire:init` to build your own lazy sort of thing.

**Medium term solution**

We could change the placeholder logic to look for the `#[Layout()]` attribute and ensure that it uses that as the layout instead of the default layout. This would ensure that the correct layout is processed from the initial page load, so it all works correctly.

Using layout inside the render method wouldn't work in this scenario (and it's possible it may never work with full page lazy components).

PRs welcome if someone wants to try this 🙂

**Longer term solution**

We're looking at better ways to handle lazy components, as such we may come across a better way to handle this when we get around to sorting them out (although I don't think there is any working around the render method part of the problem).

Hope this helps!

**Reply by joshhanley** [reply]
If someone does want to have a look at fixing the layout attribute loading, I would start looking here https://github.com/livewire/livewire/blob/main/src/Features/SupportPageComponents/HandlesPageComponents.php#L14-L20

**Reply by DanielSpravtsev** [reply]
@joshhanley Thank you for investigation into this problem. I digged it out a little bit and gone to something like that: 
We could put something like this
`store($this->component)->set('LayoutAttribute', $reflectionClass->getAttributes(Layout::class)[0]->getArguments()[0]);`
here https://github.com/livewire/livewire/blob/f1faf4899e2af8fa489cea6c43721fddc60cc2e3/src/Features/SupportLazyLoading/SupportLazyLoading.php#L52

And then replace original line to that
`$layoutConfig = $layoutConfig ?: new PageComponentConfig(view: store($this::class)->get('LayoutAttribute'));`
here
https://github.com/livewire/livewire/blob/f1faf4899e2af8fa489cea6c43721fddc60cc2e3/src/Features/SupportPageComponents/HandlesPageComponents.php#L20C66-L20C66

But second part is just a theory, because I don't know how to intercept Livewire component and `LayoutAttribute` value at HandlesPageComponents.php 

Any ideas?

**Reply by joshhanley** [reply]
@DanielSpravtsev hmm not sure exactly. I would just try to get it all running in `HandlesPageComponents` to start with, and get a proof of concept working first, and then worry about moving logic out to `SupportLazyLoading`.

I haven't had a look at the code to see how it runs though, just found the starting point to make it easier for others to have a look into, so I'm not really sure how any of it works at this stage 🙂

**Reply by gazzoy** [reply]
@joshhanley Thank you vm for yoru very detailed insights. Will take a short term workaround at this point!  Would tackle with the med term one when I have a time of course.

**Reply by gazzoy** [reply]
@joshhanley @DanielSpravtsev 

I made a PR #7551  for this, but I really don't think my fix is good enough as I'm too newbie for Livewire and feel I must missing something. Please let me know if you have any feedbacks or comments.
